expr/values.go

```diff
@@ -0,0 +1,233 @@
+package expr
+
+import (
+	"github.com/brimsec/zq/zcode"
+	"github.com/brimsec/zq/zng"
+	"github.com/brimsec/zq/zson"
+)
+
+type RecordExpr struct {
+	zctx    *zson.Context
+	typ     *zng.TypeRecord
+	builder *zcode.Builder
+	columns []zng.Column
+	exprs   []Evaluator
+}
+
+func NewRecordExpr(zctx *zson.Context, names []string, exprs []Evaluator) *RecordExpr {
+	columns := make([]zng.Column, 0, len(names))
+	for _, name := range names {
+		columns = append(columns, zng.Column{Name: name})
+	}
+	return &RecordExpr{
+		zctx:    zctx,
+		builder: zcode.NewBuilder(),
+		columns: columns,
+		exprs:   exprs,
+	}
+}
+
+func (r *RecordExpr) Eval(rec *zng.Record) (zng.Value, error) {
+	var changed bool
+	b := r.builder
+	b.Reset()
+	for k, e := range r.exprs {
+		zv, err := e.Eval(rec)
+		if err != nil {
+			return zng.Value{}, err
+		}
+		if r.columns[k].Type != zv.Type {
+			r.columns[k].Type = zv.Type
+			changed = true
+		}
+		if zng.IsContainerType(zv.Type) {
+			b.AppendContainer(zv.Bytes)
+		} else {
+			b.AppendPrimitive(zv.Bytes)
+		}
+	}
+	if changed {
+		var err error
+		r.typ, err = r.zctx.LookupTypeRecord(r.columns)
+		if err != nil {
+			return zng.Value{}, err
+		}
+	}
+	return zng.Value{r.typ, b.Bytes()}, nil
+}
+
+type ArrayExpr struct {
+	zctx    *zson.Context
+	typ     *zng.TypeArray
+	builder *zcode.Builder
+	exprs   []Evaluator
+}
+
+func NewArrayExpr(zctx *zson.Context, exprs []Evaluator) *ArrayExpr {
+	return &ArrayExpr{
+		zctx:    zctx,
+		typ:     zctx.LookupTypeArray(zng.TypeNull),
+		builder: zcode.NewBuilder(),
+		exprs:   exprs,
+	}
+}
+
+func (a *ArrayExpr) Eval(rec *zng.Record) (zng.Value, error) {
+	inner := a.typ.Type
+	container := zng.IsContainerType(inner)
+	b := a.builder
+	b.Reset()
+	var first zng.Type
+	for _, e := range a.exprs {
+		zv, err := e.Eval(rec)
+		if err != nil {
+			return zng.Value{}, err
+		}
+		typ := zv.Type
+		if first == nil {
+			first = typ
+		}
+		if typ != inner && typ != zng.TypeNull {
+			if typ == first || first == zng.TypeNull {
+				a.typ = a.zctx.LookupTypeArray(zv.Type)
+				inner = a.typ.Type
+				container = zng.IsContainerType(inner)
+			} else {
+				return zng.NewErrorf("illegal mixed type array"), nil
```

Comment:
I think this is ok for now and yes it was deliberate.  We can fold this into a more comprehensive issue about what to do for mixed type arrays because we could handle this case but also turn them in to array of unions if the mixed elements aren't type compatible.  Also, the current errors are embedded so they would shed light to the user as to the mixed-type nature of their query (instead of just stopping with a type error).  I'm on the fence as to just doing something that works or generating errors and making the user cast, e.g., `put ports=[80:port, id.resp_p, id.resp_p]`.  Also, if you have a missing value should the result be missing or be a partial results (e.g., `foo:null` for a record field that is missing).  See issue #2411